packages/@lwc/template-compiler/src/codegen/codegen.ts

```diff
@@ -751,7 +756,15 @@ export default class CodeGen {
                         isSvgUseHref(currentNode.name, name, currentNode.namespace) &&
                         !isBooleanLiteral(value);
 
-                    if (isExpression(value) || isIdOrIdRef || isSvgHref) {
+                    // `<a href="#foo">` and `<area href="#foo">` must be dynamic due to synthetic shadow scoping
+                    // Note this only applies if there is an `id` attribute somewhere in the template
+                    const isScopedFragmentRef =
+                        this.scopeFragmentId &&
+                        isStringLiteral(value) &&
+                        isAllowedFragOnlyUrlsXHTML(currentNode.name, name, currentNode.namespace) &&
+                        isFragmentOnlyUrl(value.value);
```

Comment:
This logic is analogous to the logic in the non-static-optimized code path:

https://github.com/salesforce/lwc/blob/96b9c2a1666105984ca67f792d49d3d1d631ec09/packages/%40lwc/template-compiler/src/codegen/index.ts#L438-L444